cirq/testing/circuit_compare_test.py

```diff
@@ -543,3 +554,38 @@ class NoProtocol:
     with pytest.raises(AssertionError, match='disagree'):
         cirq.testing.assert_has_consistent_qid_shape(InconsistentOp3())
     cirq.testing.assert_has_consistent_qid_shape(NoProtocol())
+
+
+def test_assert_apply_unitary_works_when_axes_transposed_failure():
+
+    class BadOp:
+
+        def _apply_unitary_(self, args: cirq.ApplyUnitaryArgs):
+            # Get a more convenient view of the data.
+            a, b = args.axes
+            rest = list(range(len(args.target_tensor.shape)))
+            rest.remove(a)
+            rest.remove(b)
+            size = args.target_tensor.size
+            view = args.target_tensor.transpose([a, b, *rest])
+            view = view.reshape((4, size // 4))  # Oops. Reshape might copy.
+
+            # Apply phase gradient.
+            view[1, ...] *= 1j
+            view[2, ...] *= -1
+            view[3, ...] *= -1j
+            return args.target_tensor
+
+        def _num_qubits_(self):
+            return 2
+
+    assert cirq.has_unitary(BadOp())
+    bad_op = BadOp()
```

Comment:
nit: no need to instantiate two, re-use.

Reply:
Done.